src/Storages/MergeTree/MergeTreeIndexGranuleBloomFilter.cpp

```diff
@@ -40,42 +24,28 @@ MergeTreeIndexGranuleBloomFilter::MergeTreeIndexGranuleBloomFilter(size_t bits_p
 }
 
 MergeTreeIndexGranuleBloomFilter::MergeTreeIndexGranuleBloomFilter(
-    size_t bits_per_row_, size_t hash_functions_, size_t total_rows_, const Blocks & granule_index_blocks_)
-        : total_rows(total_rows_), bits_per_row(bits_per_row_), hash_functions(hash_functions_)
+    size_t bits_per_row_, size_t hash_functions_, const std::vector<HashSet<UInt64>>& column_hashes_)
+        : bits_per_row(bits_per_row_), hash_functions(hash_functions_), bloom_filters(column_hashes_.size())
 {
-    if (granule_index_blocks_.empty() || !total_rows)
+    if (column_hashes_.empty())
         throw Exception("LOGICAL ERROR: granule_index_blocks empty or total_rows is zero.", ErrorCodes::LOGICAL_ERROR);
 
-    assertGranuleBlocksStructure(granule_index_blocks_);
-
-    for (size_t index = 0; index < granule_index_blocks_.size(); ++index)
-    {
-        Block granule_index_block = granule_index_blocks_[index];
+    size_t bloom_filter_max_size = 0;
+    for (const auto & column_hash : column_hashes_)
+        bloom_filter_max_size = std::max(bloom_filter_max_size, column_hash.size());
```

Comment:
Hi @nikitamikhaylov , thanks for looking at this change.

Different columns may have different data cardinality. Since I'm using a hash set (column_hashes) for each column,  only the unique hash elements are stored and the number of items in each `column_hash` may be different from another.
Because of the way the bloom filter is serialized to disk (there's only one variable for the column size), we use the highest cardinality column to fit them all. Does it make sense?